### ISSUE TITLE
`-Wunused-template` many false positives for GPU code

### ISSUE THREAD

**Issue by Jacobfaib** [llvm/llvm-project#219968]
`-Wunused-template` (added by default to `-Wall` in https://github.com/llvm/llvm-project/pull/208001), is essentially unusable for templated GPU libraries such as CCCL (https://github.com/NVIDIA/cccl).

Generic GPU libraries are often header-only and expose all of the GPU kernels in those headers. `nvcc`, however, when preprocessing kernels will generate `static` stub functions for those kernels resulting in many warnings such as:
```
cub/cub/device/dispatch/kernels/kernel_transform.cuh:1150:53: error: unused function template '__wrapper__device_stub_transform_kernel' [-Werror,-Wunused-template]
 1150 | __attribute((__visibility__("hidden"))) static void __wrapper__device_stub_transform_kernel(const Offset &
      |                                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cub/cub/device/dispatch/kernels/kernel_transform.cuh:1150:53: error: unused function template 'transform_kernel' [-Werror,-Wunused-template]
 1150 | __attribute((__visibility__("hidden"))) static void transform_kernel(const Offset
      |                                                     ^~~~~~~~~~~~~~~~
cub/cub/device/dispatch/kernels/kernel_for_each.cuh:94:53: error: unused function template '__wrapper__device_stub_dynamic_kernel' [-Werror,-Wunused-template]
   94 | __attribute((__visibility__("hidden"))) static void __wrapper__device_stub_dynamic_kernel(const OffsetT &num_items, OpT &op) { ::cudaLaunchKernel(0, 0, 0, 0, 0, 0);}
      |                                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Note the `static` does not exist in the original declarations.

```sh
$ clang --version
Homebrew clang version 23.1.0
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /home/linuxbrew/.linuxbrew/Cellar/llvm/23.1.0/bin
```

**Comment by Jacobfaib** [comment]
To be clear, I think the warning is quite useful (and indeed it has surfaced a few `static inline` functions that should have just been `inline` :)), but at the moment it's not possible to keep enabled permanently because we cannot necessarily control the output of intermediate compilers.

**Comment by shafik** [comment]
I think this is: https://github.com/llvm/llvm-project/issues/218429 which led to a revert.

CC @cor3ntin

**Comment by Jacobfaib** [comment]
Yeah, they seem pretty similar. Will (is?) the reversion being backported to 23.X releases?

**Comment by Artem-B** [comment]
Is this about using clang as the host compiler from nvcc? Or about clang used as the CUDA compiler? Or both?

**Comment by Jacobfaib** [comment]
Clang is the host compiler, nvcc is the device compiler so -ccbin=clang++. I haven't tried this with using clang as the device compiler

**Comment by boomanaiden154** [comment]
What would the timeline be like for `nvcc` to stop emitting `static` on those stub functions?

**Comment by Jacobfaib** [comment]
We raised it with NVCC folks but they consider this "a host compiler issue".

We can continue pushing for it but at present there seems to be no immediate plans to fix it.
